### PR TITLE
feature/CB2-3618: Loading spinner

### PR DESCRIPTION
**Pull request by NiallBunting**
## Ticket title

_One line description_
[link to ticket number]()

## Checklist

- [ ] Branch is rebased against the latest develop/common
- [ ] Necessary `id` required prepended with `"test-"` have been checked with automation testers and added
- [ ] Code and UI has been tested manually after the additional changes
- [ ] PR title includes the JIRA ticket number
- [ ] Squashed commits contain the JIRA ticket number
- [ ] Link to the PR added to the repo
- [ ] Delete branch after merge

### CODE REVIEW

File: src/app/layout/spinner/spinner.component.html [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<div class="overlay" *ngIf="(spinnerService.showSpinner$ | async) == true">
+  <div class="govuk-box spinner-wrapper" aria-busy="true">
+    <div id="loader" class="govuk-c-loader" role="progressbar" aria-label="Loading, please wait." tabindex="0">
+      <div class="spinner-container">
+        <div class="spinner">
+          <div></div>
```
Comment (tom-bjss) [inline src/app/layout/spinner/spinner.component.html:6]: What are all these divs for ?

Comment (NiallBunting) [reply src/app/layout/spinner/spinner.component.html:6]: Came from the old implementation, so unsure but just copied it as is.

File: src/app/layout/spinner/spinner.service.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import { select, Store } from '@ngrx/store';
+import { State } from '@store/.';
+import { spinnerState, SpinnerState } from '@store/spinner/reducers/spinner.reducer';
+import { Subject, combineLatest, map, Observable } from 'rxjs';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class SpinnerService implements OnDestroy {
+  private readonly _destroying$ = new Subject<void>();
```
Comment (danielchildsbjss) [inline src/app/layout/spinner/spinner.service.ts:11]: Is this needed?

**Review by tom-bjss — DISMISSED** [review]

File: src/app/features/global-error/global-error.component.spec.ts [hunk 3]
```diff
@@ -8,7 +8,7 @@ import { GlobalErrorService } from './global-error.service';
 
 @Component({
   selector: 'app-mock-component',
-  template: '<app-global-error></app-global-error><input id="test-input" type="text" />',
+  template: '<app-global-error></app-global-error><input id="test-input" type="text" />\n',
```
Comment (mcadenas-bjss) [inline src/app/features/global-error/global-error.component.spec.ts:11]: Thanks! Couldn't figure out how to get rid of that lint warning.

File: src/app/layout/spinner/spinner.service.spec.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    mockStore = TestBed.inject(Store);
+
+    service = new SpinnerService(mockStore);
+  });
+
+  it('should create the user service', () => {
+    expect(service).toBeTruthy();
+  });
+
+  it('should get and set the username', (done) => {
```
Comment (mcadenas-bjss) [inline src/app/layout/spinner/spinner.service.spec.ts:31]: Test name doesn't match test.

**Review by mcadenas-bjss — APPROVED** [review]

**Review by danielchildsbjss — DISMISSED** [review]

**Review by mcadenas-bjss — APPROVED** [review]

File: mock-api/mock-api-server.ts [hunk 5]
```diff
@@ -23,7 +23,23 @@ server.get('/vehicles/*', (req, res) => {
...
+          console.log('Delaying not found request');
+          setTimeout(() => {
+            res.status(500);
+            res.statusMessage = 'Unavailable';
+            res.jsonp('Error service unavailable');
+          }, 2500);
+          break;
+        default:
+          res.jsonp(mockVehicleTecnicalRecordList());
+      }
```
Comment (andrewf-bjss) [inline mock-api/mock-api-server.ts:42]: I feel like we'd started down a path of command-line switches to affect the mock-api responses, whereas this approach introduces magic search criteria. Could we stick to one approach for simplicity's sake?

Personally I like the command-line approach, but I don't mind the change if there's a compelling reason.

Staying command-line, we could wrap the entire switch statement in a `setTimeout(() => {...}, getDelay())` where `getDelay()` returns the value from the command line switch or return zero if not defined so the response is not delayed.

We could use this same approach for all of the mocked api calls (i.e. test-records), and it keeps the NotFound, and ServiceError logic to a single implementation (the code as-is has 2 different HTTP responses for NotFound - a 404 (line 16) and 500 (line 35)).

@danielchildsbjss - what do you think?

**Review by andrewf-bjss — DISMISSED** [review]

File: mock-api/mock-api-server.ts (hunk 5 above)
Comment (NiallBunting) [reply mock-api/mock-api-server.ts:42]: The reason I didn't do that is it's a pain to keep changing what you want from the api. It's much easier to leave the service running and just choose what you want. Eg. with the error message stuff with @gjulien-bjss it was much easier to just do `return404` and `return202` commands to check that the error messages were working as expected.

Comment (NiallBunting) [reply mock-api/mock-api-server.ts]: Converted the entire PR to this method, awaiting discussion.

Comment (danielchildsbjss) [reply mock-api/mock-api-server.ts]: I'm happy with this approach, but we might start hitting problems if/when we add validation to the VIN search box and these secret commands can't get passed through.
Some of these things I'm not sure we need - if we want to simulate 500 not available why not just close the mock server and make it really unavailable?

**Review by mcadenas-bjss — APPROVED** [review]

**Review by tom-bjss — APPROVED** [review]

**Pull request merged**